src/Aspire.Hosting.AWS/CloudFormation/CloudFormationExtensions.cs

```diff
@@ -0,0 +1,171 @@
+    /// CloudFormation service client to a non-standard CloudFormation endpoint like an emulator.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <param name="cloudFormationClient">The AWS CloudFormation service client.</param>
+    public static IResourceBuilder<ICloudFormationResource> WithReference(this IResourceBuilder<ICloudFormationResource> builder, IAmazonCloudFormation cloudFormationClient)
+    {
+        builder.Resource.CloudFormationClient = cloudFormationClient;
+        return builder;
+    }
+
+    /// <summary>
+    /// Add a reference of a CloudFormations stack to a project. The output parameters of the CloudFormation stack are added to the project IConfiguration.
+    /// </summary>
+    /// <typeparam name="TDestination"></typeparam>
+    /// <param name="builder"></param>
+    /// <param name="cloudFormationResourceBuilder">The CloudFormation resource.</param>
+    /// <param name="configSection">The config section in IConfiguration to add the output parameters.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<TDestination> WithReference<TDestination>(this IResourceBuilder<TDestination> builder, IResourceBuilder<ICloudFormationResource> cloudFormationResourceBuilder, string configSection = "AWS::Resources")
+        where TDestination : IResourceWithEnvironment
+    {
+        var referenceResource = builder.ApplicationBuilder.AddResource(new CloudFormationReferenceResource(cloudFormationResourceBuilder.Resource, builder.Resource));
+        referenceResource.WithManifestPublishingCallback(referenceResource.Resource.WriteToManifest);
+
+        if (cloudFormationResourceBuilder.Resource is CloudFormationResource impl)
+        {
+            impl.References.Add(referenceResource.Resource);
+        }
+
+        builder.WithEnvironment(async context =>
+        {
+            if (context.ExecutionContext.IsPublishMode)
+            {
+                return;
+            }
+
+            if (cloudFormationResourceBuilder.Resource.ProvisioningTaskCompletionSource is not null)
+            {
+                context.Logger?.LogInformation("Waiting on CloudFormation resource {Name} ...", cloudFormationResourceBuilder.Resource.Name);
+                await cloudFormationResourceBuilder.Resource.ProvisioningTaskCompletionSource.Task.WaitAsync(context.CancellationToken).ConfigureAwait(false);
+            }
+
+            if (cloudFormationResourceBuilder.Resource.Outputs == null)
+            {
+                return;
+            }
+
+            configSection = configSection.Replace(':', '_');
+
+            foreach (var output in cloudFormationResourceBuilder.Resource.Outputs)
+            {
+                var envName = $"{configSection}__{output.OutputKey}";
+                context.EnvironmentVariables[envName] = output.OutputValue;
+            }
```

Comment:
Would something like this be better? You're storing output references in the environment so that it works both in manifest mode and at runtime.
 
```suggestion
            var resource = cloudFormationResourceBuilder.Resource;
            if (resource.ProvisioningTaskCompletionSource is not null)
            {
                context.Logger?.LogInformation("Waiting on CloudFormation resource {Name} ...", resource.Name);
                await resource.ProvisioningTaskCompletionSource.Task.WaitAsync(context.CancellationToken).ConfigureAwait(false);
            }

            if (cloudFormationResourceBuilder.Resource.Outputs == null)
            {
                return;
            }

            configSection = configSection.Replace(':', '_');

            foreach (var output in cloudFormationResourceBuilder.Resource.Outputs)
            {
                var envName = $"{configSection}__{output.OutputKey}";
                context.EnvironmentVariables[envName] = new StackOutputReference(output.Name, resource);
            }
```

Reply:
That would involve running the CloudFormation APIs to get the `Stack` object which populates the `cloudFormationResourceBuilder.Resource.Outputs` when publishing the manifest. In theory I could parse the template file but I would rather avoid that. The templates can be written in either JSON or YAML so I would have to write 2 different parse logics. The logic isn't hard but YamlDotnet is spotty parsing CloudFormation templates with some of the special features of CloudFormation templates.